### ISSUE TITLE
FTP vs TFTP...

### ISSUE THREAD

**Issue by buserror** [ThorstenBr/Apple2Card#3]
Not a bug, more like an option; I have a very simple tftp implementation, I was thinking of shoehorning this in, possibly to replace the FTP implementation in the firmware, it should be a lot smaller (no tcp needed) and maps more or less 1:1 to SD card 512 block sizes...

There's some drawback of course (mostly geared for upload/download, not file browsing) but I wonder what you'd think of that... basically given that probably 95% of the use case is just uploading the (often, same named) file, TFTP should fit the bill?

**Comment by ThorstenBr** [comment]
Hmm, not sure if it's an improvement. I quite like the option of using a standard FTP client and being able to look at and manage the cards. It's even showing the ProDOS/SOS volume names of each file. Also, even if FTP is a bit outdated now, it's pretty well known. And almost everyone owning an Apple II would know FTP. I could imagine TFTP could be a hurdle for some.

But there is room for improvement anyway: in fact, currently we're adding two Wiznet/Ethernet libraries to the firmware. Dan used a stripped-down variant of the Ethernet library, for the IP-forwarding support (IP65). For FTP I needed the full-blown Ethernet library. So, now both are being added - and the low-level functions are actually duplicated. Initially I was looking into unifying this. But since eventually both actually did fit into the flash, I did not proceed with the optimization further (and it would be quite a bit of work, too).

**Comment by buserror** [comment]
Actually this is interesting, as ip65 works here, but not FTP, so there might still be a lurking issue in the base support for arduino 644 I've been using, hadn't realized (yet) they had different base access to the w5500

**Comment by ThorstenBr** [comment]
You need to configure the IP address for FTP manually. It's not using DHCP (unlike IP65).
And occasionally I'm seeing initialization issues when switching the machine on. I couldn't trace the root cause of this issue yet. When that happens, both LEDs of the Wiznet module just blink in a regular fashion (more or less alternating) and it doesn't get any connection. In this case it needs a power-cycle (simple Arduino reset doesn't help).

**Comment by buserror** [comment]
Oh good to know, yes I had figured out how to set the IP manually (I can ping it) but no ftp (nothing on port 21) -- I'll check the LEDs next, perhaps it's the same bug? Thanks for the tip! perhaps we need a wiki to collect some more documentation (You guys have done a great job with the main readme already!)

**Comment by ThorstenBr** [comment]
Also remember that FTP and IP65 are exclusive. As soon as IP65 is started, the FTP server is shut down (and the Arduino just forwards the 6502's traffic). So FTP only works before booting an IP65 image (but FTP still works while any other image was running).

**Comment by buserror** [comment]
Ok I thought i had that one figured out -- altho, I ASSUMED the ftp server would be running on the 'menu' because that's where I did my tests... (after setting the manual IP)

**Comment by ThorstenBr** [comment]
Yes, FTP will also run while the menu is active, of course. But it sometimes needs a proper power-cycle after changing/setting the IP.
